### ISSUE TITLE
Support returning Optional<?> and empty Collections instead of throwing EmptyResultDataAccessException/IncorrectResultSizeDataAccessException

### ISSUE THREAD

**Issue by ryonday** [derjust/spring-data-dynamodb#27]
The current spring-data-dynamodb throws `EmptyResultDataAccessException` or IncorrectResultSizeDataAccessException when queries fail. As these are unchecked exceptions, it really makes my jaw drop when it happens unexpectedly since most times these are recoverable errors for my codebase.

Recently, I've been switching my code over to using the `Optional<T>` monad, and Spring has [touted support for this](https://spring.io/blog/2014/05/21/what-s-new-in-spring-data-dijkstra#support-for-wrapper-types-as-return-values) in new versions of spring-data. It is much more developer friendly to return these wrapper types for single-item queries and to return empty `Collection`/`Iterable` for empty multi-item queries.

**Comment by derjust** [comment]
Hi @ryonday 
the intention is for sure to support that.
Do you have a sample application (or even better, a unit test) and a stacktrace?

The most current version [`4.2`](http://mvnrepository.com/artifact/com.github.derjust/spring-data-dynamodb/4.2.0) should support the feature as shown by [this testcase](https://github.com/derjust/spring-data-dynamodb/blob/master/src/test/java/org/socialsignin/spring/data/dynamodb/domain/sample/Jdk8IT.java).

**Comment by wooder79** [comment]
I work with @ryonday and have just run into the exception being thrown although we've specified an Optional return in the interface.

```
identity_1 | Caused by: org.springframework.dao.EmptyResultDataAccessException: No results found
identity_1 | 	at org.socialsignin.spring.data.dynamodb.query.AbstractMultipleEntityQuery.getSingleResult(AbstractMultipleEntityQuery.java:41)
identity_1 | 	at org.socialsignin.spring.data.dynamodb.repository.query.AbstractDynamoDBQuery$SingleEntityExecution.execute(AbstractDynamoDBQuery.java:255)
identity_1 | 	at org.socialsignin.spring.data.dynamodb.repository.query.AbstractDynamoDBQuery.execute(AbstractDynamoDBQuery.java:288)
identity_1 | 	at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.doInvoke(RepositoryFactorySupport.java:462)
identity_1 | 	at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.invoke(RepositoryFactorySupport.java:440)
identity_1 | 	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
identity_1 | 	at org.springframework.data.projection.DefaultMethodInvokingMethodInterceptor.invoke(DefaultMethodInvokingMethodInterceptor.java:61)
identity_1 | 	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
identity_1 | 	at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:92)
identity_1 | 	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
identity_1 | 	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:208)
identity_1 | 	at com.sun.proxy.$Proxy145.findByUsernameAndProvider(Unknown Source)
```

**Comment by derjust** [comment]
Released version`4.3.1` with this fix.
Please give it some time to sync to Maven Central
